`to_json` `temporal_mode="seconds"`/`"milliseconds"` ignores the timezone offset, producing a wrong unix timestamp for aware datetimes

### Initial Checks

- [x] I confirm that I'm using Pydantic V2

### Description

When serializing a timezone-aware `datetime` with `pydantic_core.to_json(..., temporal_mode="seconds")` (or `"milliseconds"`, or the `ser_json_temporal` config equivalent), the timezone offset is silently dropped. The wall-clock time is treated as if it were UTC, so the resulting unix timestamp is wrong by exactly the offset.

For example, `datetime(2026, 7, 9, 12, 0, 0, tzinfo=timezone(timedelta(hours=-5)))` represents the instant `2026-07-09T17:00:00Z` (unix `1783616400.0`), but `temporal_mode="seconds"` serializes it as `1783598400.0` — the same value as `12:00:00Z`. It should equal Python's `datetime.timestamp()`.

**Expected** (`temporal_mode="seconds"`):

| input | serialized seconds |
|---|---|
| `2026-07-09T12:00:00+00:00` | `1783598400.0` |
| `2026-07-09T12:00:00-05:00` | `1783616400.0` |

**Actual:**

| input | serialized seconds |
|---|---|
| `2026-07-09T12:00:00+00:00` | `1783598400.0` |
| `2026-07-09T12:00:00-05:00` | `1783598400.0` ❌ (offset dropped) |

By contrast, `temporal_mode="iso8601"` correctly preserves the offset (`"2026-07-09T12:00:00-05:00"`), so only the float/timestamp modes are affected.

**Root cause.** The float conversion sums the date's UTC-midnight timestamp with the *wall-clock* time-of-day and never applies `tz_offset` (`pydantic-core/src/serializers/type_serializers/datetime_etc.rs`

`datetime_to_milliseconds` / `time_to_milliseconds` have the same shape. `speedate`'s `Time` carries a `tz_offset`, but these functions ignore it. This has been present since the `seconds`/`milliseconds` temporal modes were introduced in [pydantic-core#1743](https://github.com/pydantic/pydantic-core/pull/1743)

### Example Code

```Python
import datetime
import pydantic_core

est = datetime.timezone(datetime.timedelta(hours=-5))
aware = datetime.datetime(2026, 7, 9, 12, 0, 0, tzinfo=est)  # == 2026-07-09T17:00:00Z

serialized = float(pydantic_core.to_json(aware, temporal_mode="seconds"))
correct = aware.timestamp()

print("pydantic seconds:", serialized)     # 1783598400.0
print("datetime.timestamp():", correct)    # 1783616400.0
print("iso8601 (correct):", pydantic_core.to_json(aware, temporal_mode="iso8601").decode())
#   "2026-07-09T12:00:00-05:00"  (offset preserved here)

assert serialized == correct, "temporal_mode='seconds' dropped the tz offset"
# AssertionError: 1783598400.0 != 1783616400.0
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.13.4
        pydantic-core version: 2.46.4
          pydantic-core build: profile=release pgo=false
               python version: 3.11.14 (main, Oct 10 2025, 12:48:11) [Clang 20.1.4 ]
                     platform: Linux-6.17.0-35-generic-x86_64-with-glibc2.39
             related packages: typing_extensions-4.16.0 mypy-2.2.0 email-validator-2.3.0 pydantic-settings-2.14.2 pydantic-extra-types-2.11.1
                       commit: unknown
```

--- a/pydantic-core/src/serializers/type_serializers/datetime_etc.rs
+++ b/pydantic-core/src/serializers/type_serializers/datetime_etc.rs
@@ -14,11 +14,11 @@ use crate::serializers::SerializationState;
 use crate::serializers::config::{FromConfig, TemporalMode};
 
 pub(crate) fn datetime_to_seconds(dt: DateTime) -> f64 {
-    dt.date.timestamp() as f64 + time_to_seconds(dt.time)
+    dt.timestamp_tz() as f64 + f64::from(dt.time.microsecond) / 1_000_000.0
 }
 
 pub(crate) fn datetime_to_milliseconds(dt: DateTime) -> f64 {
-    dt.date.timestamp_ms() as f64 + time_to_milliseconds(dt.time)
+    dt.timestamp_tz() as f64 * 1_000.0 + f64::from(dt.time.microsecond) / 1_000.0
 }
 
 pub(crate) fn date_to_seconds(date: Date) -> f64 {

--- a/pydantic-core/tests/serializers/test_datetime.py
+++ b/pydantic-core/tests/serializers/test_datetime.py
@@ -1,4 +1,5 @@
 from datetime import date, datetime, time, timedelta, timezone
+from typing import Literal
 
 import pytest
 
@@ -192,6 +193,62 @@ def test_config_datetime(
         ),
     ):
         assert s.to_json({dt: 'foo'}) == expected_to_json_dict
+
+
+@pytest.mark.parametrize(
+    ['dt', 'expected', 'expected_json', 'expected_key', 'mode'],
+    [
+        (datetime(2026, 1, 1, tzinfo=timezone.utc), 1767225600.0, b'1767225600.0', '1767225600', 'seconds'),
+        (
+            datetime(2026, 1, 1, tzinfo=timezone.utc),
+            1767225600000.0,
+            b'1767225600000.0',
+            '1767225600000',
+            'milliseconds',
+        ),
+        (datetime(2026, 1, 1, tzinfo=tz(hours=-5)), 1767243600.0, b'1767243600.0', '1767243600', 'seconds'),
+        (
+            datetime(2026, 1, 1, tzinfo=tz(hours=-5)),
+            1767243600000.0,
+            b'1767243600000.0',
+            '1767243600000',
+            'milliseconds',
+        ),
+        (datetime(2026, 1, 1, tzinfo=tz(hours=2, minutes=30)), 1767216600.0, b'1767216600.0', '1767216600', 'seconds'),
+        (
+            datetime(2026, 1, 1, 1, 1, 1, 23, tzinfo=tz(hours=-5)),
+            1767247261.000023,
+            b'1767247261.000023',
+            '1767247261.000023',
+            'seconds',
+        ),
+        (
+            datetime(2026, 1, 1, 1, 1, 1, 23, tzinfo=tz(hours=-5)),
+            1767247261000.023,
+            b'1767247261000.023',
+            '1767247261000.023',
+            'milliseconds',
+        ),
+    ],
+)
+def test_config_datetime_tz_aware(
+    dt: datetime, expected: float, expected_json: bytes, expected_key: str, mode: Literal['seconds', 'milliseconds']
+):
+    """https://github.com/pydantic/pydantic/issues/13423"""
+    if mode == 'seconds':
+        assert expected == dt.timestamp()
+
+    s = SchemaSerializer(core_schema.datetime_schema(), config={'ser_json_temporal': mode})
+    assert s.to_python(dt) == dt
+    assert s.to_python(dt, mode='json') == expected
+    assert s.to_json(dt) == expected_json
+
+    key_s = SchemaSerializer(
+        core_schema.dict_schema(core_schema.datetime_schema(), core_schema.str_schema()),
+        config={'ser_json_temporal': mode},
+    )
+    assert key_s.to_python({dt: 'foo'}, mode='json') == {expected_key: 'foo'}
+    assert key_s.to_json({dt: 'foo'}) == f'{{"{expected_key}":"foo"}}'.encode()
 
 
 @pytest.mark.parametrize(

